src/status_im2/contexts/add_new_contact/events.cljs

```diff
@@ -23,50 +18,73 @@
       (let [{:keys [error]} (types/json->clj resp)]
         (if error
           (on-error error)
-          (on-success
-           (str "0x" (subs resp 5)))))))))
+          (on-success (str "0x" (subs resp 5)))))))))
+
+(re-frame/reg-fx
+ :contacts/resolve-public-key-from-ens-name
+ (fn [{:keys [chain-id ens-name on-success on-error]}]
+   (ens/pubkey chain-id ens-name on-success on-error)))
+
+(defn fx-callbacks
+  [input ens-name]
+  {:on-success (fn [pubkey]
+                 (rf/dispatch
+                  [:contacts/set-new-identity-success
+                   input ens-name pubkey]))
+   :on-error   (fn [err]
+                 (rf/dispatch
+                  [:contacts/set-new-identity-error err input]))})
+
+(defn identify-type
```

Comment:
>The reason I extracted this as a function was to make the decision-making logic more testable and to get rid of the extra baggage of dealing with the :on-success and :on-failure

Nice approach :)

>And looking again at identify-type, it's somewhat coupled to the event's validation logic,

I see `identify-type` does generally 3 things: 1. validation as a means of matching the type of the input, 2. key/id extraction and 3. it returns a map describing the parsed input. This function really looks like some kind of parser given it transforms a raw invite URL into a data structure suitable for further processing. I don't see any coupling between this function and the event handlers, except for the fact that's in the same namespace ;)

I'm not sure if you agree with me, but `parse-invite-url` seems like a better name for it, because the function is not just identifying a type, it's parsing a specific invite URL string -> data representation.

The snippet below should be close to working, and I think it promotes a better separation of responsibilities. Notice the proposal for a new function to correctly validate an ENS name `ens.core/valid?`. I'm sure there's a lib for that too, that'd be even better than reinventing the wheel. Here I'm also suggesting the matcher for the id/key type should return `unknown` when no match is found. It could be `nil` too.

I like to sometimes code for reviewing PRs, please don't interpret this comment as me trying to tell you what you should code :relaxed: 

The original code also reminded me of a tactic for refactoring, I think first shared by Sandi Metz, called [east-oriented coding](https://github.com/btihen/East_Oriented_Code). East goes to the right in the traditional Eurocentric Mercator's map. That means reducing the amount of west-oriented (left-oriented) code. Applied to a functional language like Clojure, it could mean using less binding declarations, and more functions that compose.

```clojure
(defn id-type
  [id]
  (cond
    (validators/valid-public-key? id)     :public-key
    (validators/valid-compressed-key? id) :compressed-key
    (ens/valid? id)                       :ens-name
    :else                                 :unknown))

(defn invite-url->id
  [invite-url]
  (when-let [url (and (re-find #"^https?://join.status.im/u/.+" invite-url)
                      (try (new js/URL invite-url)
                           (catch js/Error _ nil)))]
    (let [pathname (.-pathname url)
          id       (nth (string/split pathname "/") 2)]
      id)))

(defn parse-invite-url
  [invite-url]
  (let [id   (invite-url->id invite-url)
        type (id-type id)]
    {:input    invite-url
     :id       id
     :type     type
     :ens-name (when (= type :ens-name)
                 (stateofus/ens-name-parse id))}))
```

Reply:
> > And looking again at identify-type, it's somewhat coupled to the event's validation logic,
> 
> I see `identify-type` does generally 3 things: 1. validation as a means of matching the type of the input, 2. key/id extraction and 3. it returns a map describing the parsed input. This function really looks like some kind of parser given it transforms a raw invite URL into a data structure suitable for further processing. I don't see any coupling between this function and the event handlers, except for the fact that's in the same namespace ;)

Yes, sorry I didn't mean coupling in the technical sense, probably should have used a different word, but I mean the function probably won't be used anywhere else (I might be wrong, in which case, I'm suggesting to move it when it is). I meant that it's accepting multiple types of input: compressed key, ens, and links that have compressed keys/ens embedded in them. Some reasons why I'm not in agreement about moving it is that:
- no other namespace else needs it right now
- we'd have to give it a longer name, because right now `identify-type` piggy-backs on the `add-new-contact` namespace (we understand it based on it being in that namespace)
- I'd have to open another buffer for it (haha but true)

> I'm not sure if you agree with me, but `parse-invite-url` seems like a better name for it, because the function is not just identifying a type, it's parsing a specific invite URL string -> data representation.

It accepts more than just url's though - it accepts compressed keys and ens names (that are not prefaced by url's)...

> The snippet below should be close to working, and I think it promotes a better separation of responsibilities. Notice the proposal for a new function to correctly validate an ENS name `ens.core/valid?`. I'm sure there's a lib for that too, that'd be even better than reinventing the wheel. Here I'm also suggesting the matcher for the id/key type should return `unknown` when no match is found. It could be `nil` too.

Yes, I like `ens.core/valid?`, I just didn't work that much on validation because there's another PR for it #14480

Reply:
>no other namespace else needs it right now

It's a moot point, and as we can see we won't agree. That's cool, we need to disagree more ;)

I believe the code should be designed around the proper layers and abstractions, and not if something is required one or more times. For example, if a utility function is needed to transform a datetime, we put it in the utility datetime namespace, even though it might be required only once. A function that parses a contact ID clearly is not related to state management (event handler layer), even though it's used there.

>I'd have to open another buffer for it (haha but true)

haha okay Jump in and back, `M-.` and `M-,` and don't let Emacs split buffers automatically if you didn't ask. Okay, now I'm diverging too much ;)

>It accepts more than just url's though - it accepts compressed keys and ens names (that are not prefaced by url's)...

True indeed, but `identify-type` is too general. I see you referred to this "thing" as `id`. The function could be named `parse-contact-id`. Par

Reply:
OK, well like I said earlier, I was teetering on it, so thanks for explaining your reasoning. I agree with reasoning that once I moved it out of the event handler, it now belonged to the world haha. Go forth little one!

I'll move it to utils.validators where it can join it's colleagues `valid-public-key?` and `valid-compressed-key?`.

A good result of this discussion is that I was toying with the idea of calling it `something-id`, so i'll rename it `parse-contact-id`. Thanks!

Reply:
>it now belonged to the world haha. Go forth little one!

Loved the sarcasm!

https://www.youtube.com/watch?v=0LwcvjNJTuM

Reply:
To be clear, no sarcasm intended at all (no mocking or contempt on my side that is). You and I both love to discuss the intricacies of how to build software... I'm just adding the humorous metaphor because it's kind of true - I extracted this function to make it testable, but in doing so, it joined the wider code base (the world). Just funny to think of code like that, like a baby/child (bird I guess is what you imagined), making its own out in the world, away from it's cozy protected event handler.

Reply:
@ilmotta As discussed, attempting to move this led to a series of circular dependencies. As agreed, this refactoring can happen in a separate PR.